### PR TITLE
feat: Add Bedrock Nova Sonic realtime provider implementing IRealtimeClient/IRealtimeClientSession

### PR DESCRIPTION
**Pull request by tarekgh**
## Summary

Adds a Bedrock Nova Sonic provider implementing the `Microsoft.Extensions.AI` Realtime abstractions (`IRealtimeClient` / `IRealtimeClientSession`), enabling real-time bidirectional audio conversations with AWS Bedrock Nova Sonic models through the standardized MEAI interface.

This PR updates the `Microsoft.Extensions.AI.Abstractions` dependency from `9.9.1` to `10.4.1` for the realtime types.

### New Files
- **`BedrockNovaRealtimeClient.cs`** -- `IRealtimeClient` implementation that wraps an `IAmazonBedrockRuntime` and creates realtime sessions via the Nova Sonic bidirectional streaming API. Includes a convenience constructor accepting access key, secret key, region, and optional model ID.
- **`BedrockNovaRealtimeSession.cs`** -- `IRealtimeClientSession` implementation that manages the bidirectional event stream, audio buffering, Nova Sonic protocol state machine, and function call orchestration (~1,540 lines).
- **`BedrockRealtimeClientTests.cs`** -- Client-level unit tests (construction, disposal, metadata, service resolution).
- **`BedrockRealtimeSessionTests.cs`** -- Session-level unit tests covering the full protocol surface area.

### Modified Files
- **`AmazonBedrockRuntimeExtensions.cs`** -- Added `AsIRealtimeClient()` extension method.
- **`AWSSDK.Extensions.Bedrock.MEAI.NetStandard.csproj`** -- Updated `Microsoft.Extensions.AI.Abstractions` dependency from `9.9.1` to `10.4.1`.

## Features

- **Bidirectional audio streaming** -- Full-duplex audio via Nova Sonic's `InvokeModelWithBidirectionalStreamAsync` with `BodyPublisher` pattern
- **Voice Activity Detection (VAD)** -- Nova Sonic uses built-in VAD; audio content stays open for the session with trailing silence for end-of-speech detection
- **Text conversations** -- Send text messages via `CreateConversationItem` with automatic content block management
- **Function calling** -- Full tool invocation support with inline invocation, priority queue for tool results, and proper Nova Sonic tool configuration protocol
- **Input/output transcription** -- Maps Nova Sonic `textOutput` events to `InputAudioTranscriptionCompleted` and `OutputAudioTranscriptionDelta` messages
- **Thread-safe sends** -- `SemaphoreSlim` serializes all outbound writes; priority queue bypasses normal channel for time-sensitive tool results
- **Graceful disposal** -- Race-safe dispose with proper protocol shutdown sequence (contentEnd -> promptEnd -> sessionEnd), semaphore-guarded cleanup
- **Convenience constructor** -- `BedrockNovaRealtimeClient(string accessKeyId, string secretAccessKey, string regionName, string? defaultModelId)` for simple setup
- **Owned runtime disposal** -- Convenience constructor tracks ownership and properly disposes its internally-created `AmazonBedrockRuntimeClient`
- **NetStandard 2.0 compatible** -- Targets NetStandard 2.0 with polyfills for `[Experimental]` attribute

## Usage Example

```csharp
using Amazon.BedrockRuntime;
using Microsoft.Extensions.AI;

// Simple: create with credentials and region
IRealtimeClient realtimeClient = new BedrockNovaRealtimeClient(
    accessKeyId, secretAccessKey, "us-east-1", "amazon.nova-2-sonic-v1:0");

// Or advanced: create from an existing IAmazonBedrockRuntime instance
// var runtime = new AmazonBedrockRuntimeClient(RegionEndpoint.USEast1);
// IRealtimeClient realtimeClient = new BedrockNovaRealtimeClient(runtime);

// Or use the extension method
// IRealtimeClient realtimeClient = runtime.AsIRealtimeClient();

// Define a tool for function calling
AIFunction getWeather = AIFunctionFactory.Create(
    (string location) =>
        location.ToLowerInvariant() switch
        {
            var l when l.Contains("seattle")       => $"The weather in {location} is rainy, 55F",
            var l when l.Contains("new york")      => $"The weather in {location} is cloudy, 70F",
            var l when l.Contains("san francisco") => $"The weather in {location} is foggy, 60F",
            _                                      => $"Sorry, I don't have weather data for {location}."
        },
    "GetWeather",
    "Gets the current weather for a given location");

// Configure session options
var sessionOptions = new RealtimeSessionOptions
{
    Instructions = "You are a helpful assistant.",
    Voice = "tiffany",
    Tools = [getWeather],
    InputAudioFormat = new RealtimeAudioFormat("audio/lpcm", 16000),
    OutputAudioFormat = new RealtimeAudioFormat("audio/lpcm", 24000),
};

// Create a session and start streaming
await using var session = await realtimeClient.CreateSessionAsync(sessionOptions);

// Start listening for server messages in the background
_ = Task.Run(async () =>
{
    await foreach (var message in session.GetStreamingResponseAsync(cancellationToken))
    {
        switch (message)
        {
            case OutputTextAudioRealtimeServerMessage audio
                when audio.Type == RealtimeServerMessageType.OutputAudioDelta:
                PlayAudio(audio.Audio);
                break;

            case OutputTextAudioRealtimeServerMessage text
                when text.Type == RealtimeServerMessageType.OutputTextDelta:
                Console.Write(text.Text);
                break;

            case InputAudioTranscriptionRealtimeServerMessage transcription:
                Console.WriteLine($"You said: {transcription.Transcription}");
                break;
        }
    }
});

// Send audio (e.g., from microphone)
var audioContent = new DataContent(audioBytes, "audio/lpcm");
await session.SendAsync(new InputAudioBufferAppendRealtimeClientMessage(audioContent));
await session.SendAsync(new InputAudioBufferCommitRealtimeClientMessage());
```

## Key Design Decisions

1. **BodyPublisher while(true) loop** -- Nova Sonic uses a `BodyPublisher` delegate that returns events one at a time. The provider uses a persistent loop that waits on both a normal `Channel<T>` for audio events and a `ConcurrentQueue` for priority tool results. This avoids closing the outbound stream prematurely while ensuring tool results bypass queued audio.

2. **Priority queue for tool results** -- Tool results are sent via `WritePriorityEvent` which enqueues to a `ConcurrentQueue` and signals the BodyPublisher via `TaskCompletionSource`. This ensures tool results reach Nova Sonic before it commits to a speculative fallback response, which is critical for reliable function calling.

3. **Inline tool invocation** -- When `FunctionInvokingRealtimeSession` middleware sends `CreateConversationItem` with tool results, the provider sends them inline via the priority queue with proper Nova Sonic protocol framing (contentStart -> toolResult -> contentEnd -> silence nudge).

4. **Audio content stays open** -- Following the official AWS Nova Sonic samples, the audio content block is never explicitly closed mid-conversation. Instead, trailing silence is sent on `InputAudioBufferCommit` to help VAD detect end-of-speech. This matches how the Nova Sonic service expects continuous audio streams.

5. **Prompt lifecycle management** -- The provider automatically manages promptStart/promptEnd lifecycle. After a tool result cycle, a new prompt is opened with a fresh `promptName` for the next turn, maintaining correct protocol state.

6. **BodyPublisher shutdown safety** -- The while(true) loop checks for cancellation token and channel completion to avoid CPU-burning spin loops during disposal. All exit paths dispose the enumerator.

7. **Thread safety** -- All outbound writes go through a `SemaphoreSlim`. Fields mutated under the semaphore (`_promptName`, `_audioContentName`) are captured to local variables when read outside the semaphore (e.g., in `SendToolResultInline`).

## Test Coverage

54 unit tests covering:
- Client lifecycle (construction, convenience constructor, disposal, owned runtime disposal)
- Session lifecycle (connect, dispose, idempotent dispose, service resolution)
- Protocol events (sessionStart, promptStart, contentStart, audioInput, contentEnd, promptEnd, sessionEnd)
- Audio streaming (append, commit with trailing silence, CreateResponse silence nudge)
- Text content (send text via CreateConversationItem)
- Function calling (tool configuration in promptStart, tool result protocol, inline tool result events)
- Error handling (dispose during active session, concurrent operations, invalid state)
- Message mapping (all server message types, usage details, completion events)
- Edge cases (null options, empty audio, multiple tool calls)

### CODE REVIEW

File: extensions/test/BedrockMEAIRealtimeTests/BedrockRealtimeClientTests.cs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+ *  http://aws.amazon.com/apache2.0
+ * 
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+using System;
+using System.IO;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/test/BedrockMEAIRealtimeTests/BedrockRealtimeClientTests.cs]: Done.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+ * permissions and limitations under the License.
+ */
+
+#if NET8_0_OR_GREATER
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.BedrockRuntime.Model;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: Done.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    private readonly IAmazonBedrockRuntime _runtime;
+    private readonly string _defaultModelId;
+    private readonly ChatClientMetadata _metadata;
+    private readonly bool _ownsRuntime;
+
+    /// <summary>Initializes a new instance of the <see cref="BedrockNovaRealtimeClient"/> class.</summary>
+    /// <param name="runtime">The Amazon Bedrock Runtime client.</param>
+    /// <param name="defaultModelId">
+    /// The default model ID. If <see langword="null"/>, defaults to <c>amazon.nova-sonic-v1:0</c>.
+    /// </param>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs:41]: done.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        _metadata = new(AmazonBedrockRuntimeExtensions.ProviderName, defaultModelId: _defaultModelId);
+    }
+
+    /// <summary>Initializes a new instance using AWS access keys and region.</summary>
+    /// <param name="accessKeyId">The AWS access key ID.</param>
+    /// <param name="secretAccessKey">The AWS secret access key.</param>
+    /// <param name="regionName">The AWS region (e.g. "us-east-1").</param>
+    /// <param name="defaultModelId">
+    /// The default model ID. If <see langword="null"/>, defaults to <c>amazon.nova-sonic-v1:0</c>.
+    /// </param>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: done

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs [hunk 5]
```diff
@@ -25,6 +25,21 @@ public static class AmazonBedrockRuntimeExtensions
     /// <summary>The provider name to use in metadata.</summary>
     internal const string ProviderName = "aws.bedrock";
 
+#if NET8_0_OR_GREATER
+    /// <summary>Gets an <see cref="IRealtimeClient"/> for the specified <see cref="IAmazonBedrockRuntime"/> instance.</summary>
+    /// <param name="runtime">The runtime instance to be represented as an <see cref="IRealtimeClient"/>.</param>
+    /// <param name="defaultModelId">
+    /// The default model ID to use when no model is specified in session options. If not specified,
+    /// defaults to <c>amazon.nova-sonic-v1:0</c>.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs]: done.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AWSSDK.Extensions.Bedrock.MEAI.NetStandard.csproj [hunk 6]
```diff
@@ -41,7 +42,7 @@
   </Choose>
   
   <ItemGroup>
-  <PackageReference Include="Microsoft.Extensions.AI.Abstractions" Version="9.9.1" />
+    <PackageReference Include="Microsoft.Extensions.AI.Abstractions" Version="10.4.1" />
   </ItemGroup>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AWSSDK.Extensions.Bedrock.MEAI.NetStandard.csproj:51]: Added generator/.DevConfigs/aec9fb2f-be53-4fc5-8ac3-ec2e7395e087.json

File: extensions/test/BedrockMEAITests/BedrockRealtimeSessionTests.cs [hunk 7]
```diff
@@ -0,0 +1,1082 @@
...
+        var options = new RealtimeSessionOptions
+        {
+            Instructions = "You are a helpful assistant.",
+            Voice = "tiffany",
+            MaxOutputTokens = 2048
+        };
+
+        await session.SendAsync(new SessionUpdateRealtimeClientMessage(options));
+        await Task.Delay(100); // Allow publisher to consume
+
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/test/BedrockMEAITests/BedrockRealtimeSessionTests.cs:331]: Fxied.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AWSSDK.Extensions.Bedrock.MEAI.NetStandard.csproj [hunk 8]
```diff
@@ -17,6 +17,7 @@
     <LangVersion>Latest</LangVersion>
     <Nullable>enable</Nullable>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <NoWarn>$(NoWarn);MEAI001;IL2026</NoWarn>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AWSSDK.Extensions.Bedrock.MEAI.NetStandard.csproj]: Switched to use source generator and removed the global `IL2026`

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs [hunk 9]
```diff
@@ -0,0 +1,1542 @@
...
+                            if (_outboundChannel.Reader.TryRead(out var part))
+                            {
+                                return part;
+                            }
+                        }
+
+                        // Check if session is shutting down (avoids spin loop when
+                        // WaitToReadAsync returns a faulted/cancelled task that
+                        // Task.WhenAny doesn't propagate)
+                        sessionToken.ThrowIfCancellationRequested();
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs:169]: The TCS is not truly orphaned in a way that causes issues here. The `WhenAny` at line 129 pairs the TCS with `normalReady` (the channel reader), so the loop never blocks on the TCS alone. When the next normal channel write arrives or the session shuts down (via `_sessionCts` cancellation), `WhenAny` completes regardless, and the TCS becomes unreferenced and GC'd.

Regarding the suggested `CancellationTokenSource` per iteration; this would add allocation and disposal overhead on every loop iteration (CTS is IDisposable and allocates internally) for a scenario that's already handled. The `WaitToReadAsync` waiter is lightweight and gets completed on the next channel write or channel completion, so there's no meaningful accumulation.

The session lifetime is bounded (single conversation), priority events are infrequent (only tool results), and shutdown already propagates cleanly through the session cancellation token. The cost of the defensive fix (per-iteration CTS allocation + disposal) outweighs the theoretical risk here.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs [hunk 10]
```diff
@@ -25,6 +25,21 @@ public static class AmazonBedrockRuntimeExtensions
     /// <summary>The provider name to use in metadata.</summary>
     internal const string ProviderName = "aws.bedrock";
 
+#if NET8_0_OR_GREATER
+    /// <summary>Gets an <see cref="IRealtimeClient"/> for the specified <see cref="IAmazonBedrockRuntime"/> instance.</summary>
+    /// <param name="runtime">The runtime instance to be represented as an <see cref="IRealtimeClient"/>.</param>
+    /// <param name="defaultModelId">
+    /// The default model ID to use when no model is specified in session options. If not specified,
+    /// defaults to <c>amazon.nova-2-sonic-v1:0</c>.
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs]: We should not default to a specific model, it will get awkward when newer models come along and when do we choose when we can update the default. Following the pattern of `AsIChatClient` if the `defaultModelId` is not set then users must set the model in the `RealtimeSessionOptions`.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        _defaultModelId = defaultModelId ?? DefaultNovaSonicModelId;
+        _metadata = new(AmazonBedrockRuntimeExtensions.ProviderName, defaultModelId: _defaultModelId);
+    }
+
+    /// <summary>Initializes a new instance using AWS access keys and region.</summary>
+    /// <param name="accessKeyId">The AWS access key ID.</param>
+    /// <param name="secretAccessKey">The AWS secret access key.</param>
+    /// <param name="regionName">The AWS region (e.g. "us-east-1").</param>
+    /// <param name="defaultModelId">
+    /// The default model ID. If <see langword="null"/>, defaults to <c>amazon.nova-2-sonic-v1:0</c>.
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: Same comment about default model

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+    private readonly IAmazonBedrockRuntime _runtime;
+    private readonly string _defaultModelId;
+    private readonly ChatClientMetadata _metadata;
+    private readonly bool _ownsRuntime;
+
+    /// <summary>Initializes a new instance of the <see cref="BedrockNovaRealtimeClient"/> class.</summary>
+    /// <param name="runtime">The Amazon Bedrock Runtime client.</param>
+    /// <param name="defaultModelId">
+    /// The default model ID. If <see langword="null"/>, defaults to <c>amazon.nova-2-sonic-v1:0</c>.
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: Same comment about default model

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    /// <param name="accessKeyId">The AWS access key ID.</param>
+    /// <param name="secretAccessKey">The AWS secret access key.</param>
+    /// <param name="regionName">The AWS region (e.g. "us-east-1").</param>
+    /// <param name="defaultModelId">
+    /// The default model ID. If <see langword="null"/>, defaults to <c>amazon.nova-2-sonic-v1:0</c>.
+    /// </param>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="accessKeyId"/>, <paramref name="secretAccessKey"/>, or <paramref name="regionName"/> is <see langword="null"/> or empty.
+    /// </exception>
+    public BedrockNovaRealtimeClient(string accessKeyId, string secretAccessKey, string regionName, string? defaultModelId = null)
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: Do we need this constructor? Following the pattern of `BedrockChatClient` it only has the constructor that takes in the `IAmazonBedrockRuntime` instance. If there is a reason I haven't seen yet then instead of taking in accessKeyId and secret key you should take in the `AWSCredentials` object. That will cover more types of ways to provide credentials.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /// <inheritdoc />
+    public async Task<IRealtimeClientSession> CreateSessionAsync(
+        RealtimeSessionOptions? options = null,
+        CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        string modelId = options?.Model ?? _defaultModelId;
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: Since I said we don't want to have a hard coded default you should throw an exception if `modelId` resolves to null.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs [hunk 15]
```diff
@@ -0,0 +1,1729 @@
...
+        string? currentContentId = null;
+        string? currentContentType = null;
+        string? currentRole = null;
+        var emittedAssistantTranscripts = new HashSet<string>(StringComparer.Ordinal);
+        int outputIndex = 0;
+        bool inResponse = false;
+        UsageDetails? latestUsage = null;
+        List<AIContent>? pendingToolCallContents = null;
+
+        _streamResponse.Body.ChunkReceived += (_, _) => { };
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: What is the reason for the dummy callbacks? My memory is if you are using the `GetAsyncEnumerator` you can ignore all the event properties.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs [hunk 16]
```diff
@@ -0,0 +1,1729 @@
...
+                    IDictionary<string, object?>? toolArguments = null;
+                    if (toolUse.TryGetProperty("content", out var tcc))
+                    {
+                        if (tcc.ValueKind == JsonValueKind.Object)
+                        {
+                            try
+                            {
+
+#pragma warning disable IL2026 // Tool arguments are dynamic JSON
+                                toolArguments = JsonSerializer.Deserialize<Dictionary<string, object?>>(tcc.GetRawText());
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: Maybe I'm missing something but this call to `Deserilize` doesn't seem safe to ignore the Native AOT warning. Can we use `Dictionary<string, JsonElement>` instead and then do what over conversions we need to the write object?

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs [hunk 17]
```diff
@@ -0,0 +1,1729 @@
...
+        if (result is JsonElement jsonElement)
+        {
+            if (jsonElement.ValueKind == JsonValueKind.Object)
+            {
+                return jsonElement.GetRawText();
+            }
+
+            // Non-object JsonElement (string, number, array, etc.) — wrap in an object
+#pragma warning disable IL2026 // Tool result values are dynamic
+            return JsonSerializer.Serialize(new NovaSonicToolResultWrapper { Result = jsonElement }, NovaSonicJsonOptions.Reflection);
```
Comment (normj) [inline extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: By doing this are we making his API and the callers of this API non Native AOT complaint? By ignoring the runtime warning we are setting ourselves up for potential error of types needed being trimmed.

File: generator/.DevConfigs/aec9fb2f-be53-4fc5-8ac3-ec2e7395e087.json [hunk 18]
```diff
@@ -0,0 +1,11 @@
+{
+  "extensions": [
+    {
+      "extensionName": "Extensions.Bedrock.MEAI",
+      "type": "minor",
+      "changeLogMessages": [
+        "Add Bedrock Nova Sonic realtime provider implementing `IRealtimeClient` / `IRealtimeClientSession` for real-time bidirectional audio conversations with tool calling support."
```
Comment (normj) [inline generator/.DevConfigs/aec9fb2f-be53-4fc5-8ac3-ec2e7395e087.json]: Can you add the `(Preview)` prefix since the APIs are marked as experimental?

**Review by normj — COMMENTED** [review]
Looks like a cool feature. I took a first pass. After you make the changes I'll try it out in action.

**Comment by tarekgh** [comment]
> Looks like a cool feature. I took a first pass. After you make the changes I'll try it out in action.

Thanks @normj for helping with the review. Just to let you know, I have a demo test app I am using to test with different providers. https://github.com/tarekgh/RealtimeProposalDemoApp.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs (hunk 10 above)
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/AmazonBedrockRuntimeExtensions.cs]: Done.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs (hunk 13 above)
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeClient.cs]: I removed the constructor for now. We can introduce one later takes AWSCredentials if we want to.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs (hunk 15 above)
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: Good catch, it was not intention to keep these. I removed them.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs (hunk 16 above)
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: Removed all IL2026 exceptions. We should be good now with building with AOT.

File: extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs (hunk 17 above)
Comment (tarekgh) [reply extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockNovaRealtimeSession.cs]: fixed.

File: generator/.DevConfigs/aec9fb2f-be53-4fc5-8ac3-ec2e7395e087.json (hunk 18 above)
Comment (tarekgh) [reply generator/.DevConfigs/aec9fb2f-be53-4fc5-8ac3-ec2e7395e087.json]: done

**Comment by tarekgh** [comment]
Hi @normj, just wanted to see if you’ve had a chance to test out that provider yet? I’m looking to wrap up this review and would love to get your take.

**Comment by normj** [comment]
@tarekgh it is on my list to hopefully get to soon.

**Review by normj — DISMISSED** [review]
I rebased the branch on the latest `development` changes and pushed a commit with a couple very minor nits. PR looks good and I did try out the sample app which was cool.

@dscpinheiro Can you do a second pass on the PR?

We are also working through some infrastructure issues that get triggered by updating the version of `Microsoft.Extensions.AI.Abstractions` which pulls in a newer version of `System.Text.Json` in some of our other process. The change to update `Microsoft.Extensions.AI.Abstractions` is fine but there are assumptions in other parts of our build system for packaging up the SDK for non NuGet users that go awry. Bare with us even when we approve the PR it might be a bit before we can merge it till we sort out the issue.

**Comment by tarekgh** [comment]
Thanks @normj!

Take your time. I want to ensure the changes is not going to cause any issue in general.

**Comment by dscpinheiro** [comment]
@tarekgh The infrastructure issues Norm mentioned have been resolved, so could you rebase your PR to get the latest changes? (we re-introduced https://github.com/aws/aws-sdk-net/pull/4184 so the latest `development` branch is already using version 10.x of `Microsoft.Extensions.AI.Abstractions`)

**Comment by tarekgh** [comment]
> @tarekgh The infrastructure issues Norm mentioned have been resolved, so could you rebase your PR to get the latest changes? (we re-introduced #4184 so the latest `development` branch is already using version 10.x of `Microsoft.Extensions.AI.Abstractions`)

I'll try to get into this soon.

**Comment by tarekgh** [comment]
@dscpinheiro updated. Let me know if you have any more asks.

**Comment by dscpinheiro** [comment]
I pushed another small commit to get the tests passing in our internal build system, but I'll merge this once I get the approvals. (In the commit I also consolidated the new test project into the existing one, and had it multi-target .NET 4.7 and .NET 8 - so we can just re-use it in the future).

**Comment by tarekgh** [comment]
The commit `7a7847f` looks good overall — it cleanly addresses the three CodeBuild issues (nuspec version, InternalsVisibleTo key, test project consolidation). A few potential concerns to discuss:

1. **Missing `InternalsVisibleTo` in NetFramework csproj** — The key was added to the NetStandard csproj but not `AWSSDK.Extensions.Bedrock.MEAI.NetFramework.csproj`. This is likely fine since the realtime tests use `#if NET8_0_OR_GREATER`, but worth confirming no net472 test path needs internals access.

2. **Missing `NoWarn` for MEAI001 in NetFramework csproj** — The NetStandard csproj suppresses `MEAI001` but the NetFramework csproj doesn't. If any MEAI experimental APIs are referenced in code compiled for net472, this could cause build failures with `TreatWarningsAsErrors` enabled.

3. **Transitive dependency on BedrockRuntime in tests** — The test project only directly references Core and the MEAI extension, relying on transitive references for `Amazon.BedrockRuntime.Model` types. This works today but could be fragile if transitive references are ever trimmed.

None of these are blockers — just items to keep in mind.

**Review by dscpinheiro — APPROVED** [review]

**Review by normj — APPROVED** [review]

**Comment by tarekgh** [comment]
Thanks @normj and @dscpinheiro for helping with this PR. Any idea when this will get published?

**Comment by dscpinheiro** [comment]
It'll be included in the next release of the AWS SDK, which happens every week day. 

The exact version of the extensions package will be in https://github.com/aws/aws-sdk-net/releases

**Comment by normj** [comment]
@tarekgh Thanks for the feature!

**Comment by dscpinheiro** [comment]
Released in https://github.com/aws/aws-sdk-net/releases/tag/4.0.245.0 (and https://www.nuget.org/packages/AWSSDK.Extensions.Bedrock.MEAI/4.0.8)
